src/gp_doctest.cpp

```diff
@@ -18,191 +18,78 @@ GPInstance MakeHelloGPInstance() {
-      EvolveLeafward{PLV::rhat_saturn, PLV::r_s, HelloGPCSP::saturn},
-      // Calculate likelihood at saturn.
-      Likelihood{HelloGPCSP::saturn, PLV::rhat_saturn, PLV::p_saturn},
-  };
+  CHECK_LT(fabs(engine->GetLogMarginalLikelihood() - -84.77961943), 1e-6);
+}
 
-  auto inst = MakeHelloGPInstance();
+TEST_CASE("GPInstance: marginal likelihood calculation") {
+  auto inst = MakeHelloGPInstanceTwoTrees();
   auto engine = inst.GetEngine();
 
-  // Test that the two-pass likelihood computation does the right thing at each internal
-  // node.
-  engine->ProcessOperations(two_pass_likelihood_computation);
-  for (size_t idx = 0; idx <= root; idx++) {
-    CHECK_LT(fabs(engine->GetLogLikelihoods()(idx) - -84.77961943), 1e-6);
-  }
+  inst.PopulatePLVs();
+  inst.ComputeLikelihoods();
+  std::cout << engine->GetLogMarginalLikelihood() << std::endl;
+  inst.EstimateBranchLengths(1e-6, 100);
 
-  // Test of our log likelihood derivative code on the jupiter branch.
-  auto jupiter_optimization = OptimizeRootward{PLV::phat_ttilde, PLV::p_jupiter,
-                                               PLV::r_ttilde, HelloGPCSP::jupiter};
-  EigenVectorXd branch_lengths = engine->GetBranchLengths();
-  auto [original_log_likelihood, log_likelihood_derivative] =
-      engine->LogLikelihoodAndDerivative(jupiter_optimization);
-  double branch_length_difference = 1e-7;
-  branch_lengths(0) += branch_length_difference;
-  engine->SetBranchLengths(branch_lengths);
-  engine->ProcessOperations(two_pass_likelihood_computation);
-  auto log_likelihood = engine->GetLogLikelihoods()(0);
-  auto derivative_estimate =
-      (log_likelihood - original_log_likelihood) / branch_length_difference;
-  CHECK_LT(fabs(derivative_estimate - log_likelihood_derivative), 1e-6);
+  CHECK_LT(fabs(engine->GetLogMarginalLikelihood() - -79.9944001), 1e-6);
+}
+
+TEST_CASE("GPInstance: gradient calculation") {
+  auto inst = MakeHelloGPInstanceSingleNucleotide();
+  auto engine = inst.GetEngine();
 
-  // Trying out optimization.
-  for (size_t pass_idx = 0; pass_idx < 8; ++pass_idx) {
-    engine->ProcessOperations(two_pass_optimization);
-  }
-  std::cout << engine->GetBranchLengths() << std::endl;
-  std::cout << engine->GetLogLikelihoods() << std::endl;
+  inst.PopulatePLVs();
+  inst.ComputeLikelihoods();
+  
+  size_t root_idx = root;
+  size_t child_idx = jupiter;
+  size_t hello_node_count = 5;
+  size_t leafward_idx = GetPlvIndex(PlvType::P,
+                                    hello_node_count,
+                                    child_idx);
+  size_t rootward_idx = GetPlvIndex(PlvType::R,
+                                    hello_node_count,
+                                    root_idx);
+  size_t pcsp_idx = inst.GetPCSPIndex(root_idx, child_idx, false);
+  OptimizeBranchLength op{leafward_idx, rootward_idx, pcsp_idx};
+  DoublePair log_lik_and_derivative = engine->LogLikelihoodAndDerivative(op);
+  // Expect log lik: -4.806671945.
```

Comment:
Expected based on what?

Reply:
Manual calculation. Can include R script that I used for calculating if you want to see how I did the calculation.